src/least_squares.rs

```diff
@@ -0,0 +1,780 @@
+    D: DataMut<Elem = E>,
+{
+    /// Solve a least squares problem of the form `Ax = rhs`
+    /// by calling `A.least_squares(rhs)`, where `rhs` is a
+    /// matrix. `A` and `rhs` are overwritten in the call.
+    ///
+    /// `A` and `rhs` must have the same layout, i.e. they must
+    /// be both either row- or column-major format, otherwise a
+    /// `IncompatibleShape` error is raised.
+    fn least_squares_in_place(
+        &mut self,
+        rhs: &mut ArrayBase<D, Ix2>,
+    ) -> Result<LeastSquaresResult<E, Ix2>> {
+        let (m, n) = (self.shape()[0], self.shape()[1]);
+        if n > m {
+            // we need a new rhs b/c it will be overwritten with the solution
+            // for which we need `n` entries
+            let k = rhs.shape()[1];
+            let mut new_rhs = Array2::<E>::zeros((n, k));
+            new_rhs.slice_mut(s![0..m, ..]).assign(rhs);
+            compute_least_squares_nrhs(self, &mut new_rhs)
+        } else {
+            compute_least_squares_nrhs(self, rhs)
+        }
+    }
+}
+
+fn compute_least_squares_nrhs<E, D1, D2>(
+    a: &mut ArrayBase<D1, Ix2>,
+    rhs: &mut ArrayBase<D2, Ix2>,
+) -> Result<LeastSquaresResult<E, Ix2>>
+where
+    E: Scalar + Lapack + LeastSquaresSvdDivideConquer_,
+    D1: DataMut<Elem = E>,
+    D2: DataMut<Elem = E>,
+{
+    let a_layout = a.layout()?;
+    let rhs_layout = rhs.layout()?;
+    let LeastSquaresOutput::<E> {
+        singular_values,
+        rank,
+    } = unsafe {
+        <E as LeastSquaresSvdDivideConquer_>::least_squares_nrhs(
+            a_layout,
+            a.as_allocated_mut()?,
+            rhs_layout,
+            rhs.as_allocated_mut()?,
+        )?
+    };
+
+    let solution: Array2<E> = rhs.slice(s![..a.shape()[1], ..]).to_owned();
+    let singular_values = Array::from_shape_vec((singular_values.len(),), singular_values)?;
+    let (m, n) = (a.shape()[0], a.shape()[1]);
+    let residual_sum_of_squares = compute_residual_array1(m, n, rank, &rhs);
+    Ok(LeastSquaresResult {
+        solution,
+        singular_values,
+        rank,
+        residual_sum_of_squares,
+    })
+}
+
+fn compute_residual_array1<E: Scalar, D: Data<Elem = E>>(
+    m: usize,
+    n: usize,
+    rank: i32,
+    b: &ArrayBase<D, Ix2>,
+) -> Option<Array1<E::Real>> {
+    if m < n || n != rank as usize {
+        return None;
+    }
+    Some(
+        b.slice(s![n.., ..])
+            .mapv(|x| x.powi(2).abs())
+            .sum_axis(Axis(0)),
+    )
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use approx::AbsDiffEq;
+    use ndarray::{ArcArray1, ArcArray2, Array1, Array2, CowArray};
+    use num_complex::Complex;
+
+    ///////////////////////////////////////////////////////////////////////////
+    /// Test cases taken from the scipy test suite for the scipy lstsq function
+    /// https://github.com/scipy/scipy/blob/v1.4.1/scipy/linalg/tests/test_basic.py
+    ///////////////////////////////////////////////////////////////////////////
```

Comment:
```suggestion
    //
    // Test cases taken from the scipy test suite for the scipy lstsq function
    // https://github.com/scipy/scipy/blob/v1.4.1/scipy/linalg/tests/test_basic.py
    //

Reply:
Please use `//` style non-doc comment here. `///` means this comment is for the next `#[test]` function.
https://doc.rust-lang.org/reference/comments.html

Reply:
Fixed.